XenModel/Network/Heartbeat.cs

```diff
@@ -148,6 +148,23 @@ private void DoHeartbeat()
                 HandleConnectionLoss();
                 return;
             }
+            catch (WebException exn)
+            {
+                log.Error(exn);
+                if (((HttpWebResponse)exn.Response).StatusCode == HttpStatusCode.ProxyAuthenticationRequired) // work-around for CA-214653
+                {
+                    log.DebugFormat("Heartbeat for {0} has failed due to {1} credentials; closing the main connection",
+                        session == null ? "null" : session.Url,
+                        session.proxy.Proxy.Credentials == null ? "missing" : "incorrect");

```

Comment:
If session==null you'll have trouble with argument {1}.
Also, although we currently don't localise log messages, it is generally not good practice to construct strings like this. Better provide the whole sentence, e.g.
if (session ==null) log.DebugFormat(....)
else if (session.proxy.Proxy.Credentials == null) log.DebugFormat(...)
else ...